drivers/rn2xx3/rn2xx3_internal.c

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright (C) 2017 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_rn2xx3
+ * @{
+ *
+ * @file
+ * @brief       Internal driver implementation for the RN2483/RN2903 devices
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "fmt.h"
+#include "rn2xx3_internal.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define RESP_TIMEOUT_SEC            (5U)
+
+static const char *closing_seq = "\r\n";
+
+static void _uart_write_str(rn2xx3_t *dev, const char *str)
+{
+    size_t len = strlen(str);
+    if (len) {
+        uart_write(dev->p.uart, (uint8_t *)str, len);
+    }
+}
+
+static void isr_resp_timeout(void *arg)
+{
+    rn2xx3_t *dev = (rn2xx3_t *)arg;
+    mutex_unlock(&(dev->resp_lock));
+}
+
+static bool _wait_reply(rn2xx3_t *dev, uint8_t timeout)
+{
+    dev->resp_done = 0;
+    dev->resp_size = 0;
+    dev->resp_buf[0] = 0;
+
+    xtimer_ticks64_t sent_time = xtimer_now64();
+
+    xtimer_t resp_timer;
+    resp_timer.callback = isr_resp_timeout;
+    resp_timer.arg = dev;
+
+    xtimer_set(&resp_timer, (uint32_t)timeout * US_PER_SEC);
+
+    /* wait for results */
+    while ((!dev->resp_done) &&
+           xtimer_less(xtimer_diff32_64(xtimer_now64(), sent_time),
+                                        xtimer_ticks_from_usec(timeout * US_PER_SEC))) {
+        mutex_lock(&(dev->resp_lock));
+    }
+
+    xtimer_remove(&resp_timer);
+
+    if (dev->resp_done == 0) {
+        DEBUG("[rn2xx3] response timeout\n");
+        return true;
+    }
+
+    return false;
+}
+
+void rn2xx3_hex_to_bytes(const char *hex, uint8_t *byte_array)
```

Comment:
no fmt here?

Reply:
The function in fmt doesn't work when it's called within the UART cb (interrupt). I had a hard time debugging this, only my initial implementation works.
If you have a good suggestion, I buy it.